### PR TITLE
Add C++/Java Solution to Bipartiteness (Silver)

### PR DESCRIPTION
**Pull request by ryanchou-dev**
sans pls go easy on the java code 🙏 

---
_If any of the below doesn't apply to this Pull Request, mark the checkbox as
done._

- [ ] I have tested my code
- [ ] I have followed the code style guidelines mentioned
      [here](https://usaco.guide/general/adding-solution/#code-conventions)
- [ ] I have properly formatted the files according to the steps mentioned
      [here](https://usaco.guide/general/adding-solution#steps)

### CODE REVIEW

File: solutions/cf-862B.mdx [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+	public static void main(String[] args) throws IOException{
+		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
+		PrintWriter out = new PrintWriter(System.out);
+
+		int n = Integer.parseInt(br.readLine());
+
+		// +1 for 1-indexed nodes
+		adj = new ArrayList[n + 1];
+		
+		for (int i = 0; i <= n; i++) adj[i] = new ArrayList<Integer>();
```
Comment (Navigator365) [inline solutions/cf-862B.mdx:44]: Just write it in the bracket style

File: solutions/cf-862B.mdx [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+			int u = Integer.parseInt(s.nextToken());
+			int v = Integer.parseInt(s.nextToken());
+			
+			adj[u].add(v);
+			adj[v].add(u);
+		}
+		
+		dfs(1, 0, false);
+		
+		long ans = (cnt[0] * cnt[1]) - (n - 1);
```
Comment (Navigator365) [inline solutions/cf-862B.mdx:58]: Maybe a comment here explaining this

**Review by Navigator365 — CHANGES_REQUESTED** [review]
Great! Just a couple things

File: solutions/cf-862B.mdx [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+```java
+import java.util.*;
+import java.io.*;
+
+public class Bipartiteness {
+	public static ArrayList<Integer> adj[];
+	
+	// counters for colored and uncolored nodes
+	public static long[] cnt = {0, 0};
```
Comment (SansPapyrus683) [inline solutions/cf-862B.mdx:33]: ryan
for those of us who have a dirty mind
please rename `cnt` to something else

File: solutions/cf-862B.mdx [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+		PrintWriter out = new PrintWriter(System.out);
+
+		int n = Integer.parseInt(br.readLine());
+
+		// +1 for 1-indexed nodes
+		adj = new ArrayList[n + 1];
+		
+		for (int i = 0; i <= n; i++) adj[i] = new ArrayList<Integer>();
+
+		StringTokenizer s;
```
Comment (SansPapyrus683) [inline solutions/cf-862B.mdx:46]: why not put this in the for loop
does it make it tle or smth

File: solutions/cf-862B.mdx [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+<CPPSection>
+
+```cpp
+#include <iostream>
+#include <vector>
+using namespace std;
+
+const int MAX_N = 100000;
+
+// +1 for 1-indexed nodes
```
Comment (SansPapyrus683) [inline solutions/cf-862B.mdx:87]: i'd ask you to make it 0-indexed but you made it so explicit and so nice that i'm just going to leave these be

File: solutions/cf-862B.mdx [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+vector<vector<int>> adj(MAX_N + 1);
+
+// counters for colored and uncolored nodes
+long long c = 0, uc = 0;
+
+void dfs(int node, int parent, bool col) {
+	if (col) c++;
+	else uc++;
```
Comment (SansPapyrus683) [inline solutions/cf-862B.mdx:95]: braces? or even better
```suggestion
	(col ? c : uc)++;
```
not sure if that works tho

File: solutions/cf-862B.mdx [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+	for (int i = 0; i < n - 1; i++) {
+		int u, v;
+		cin >> u >> v;
+		adj[u].push_back(v);
+		adj[v].push_back(u);
+	}
+
+	dfs(1, 0, 0);
+
+	cout << (long long) (c * uc) - (n - 1) << "\n";
```
Comment (SansPapyrus683) [inline solutions/cf-862B.mdx:117]: maybe an explanation for this in a comment?

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/cf-862B.mdx (hunk 4 above)
Comment (ryanchou-dev) [reply solutions/cf-862B.mdx:46]: it's faster, does it make more sense if it is in the for loop?

Comment (ryanchou-dev) [reply solutions/cf-862B.mdx:46]: resolved by using kattio

**Comment by ryanchou-dev** [comment]
thanks for the reviews!

File: solutions/cf-862B.mdx (hunk 7 above)
Comment (bqi343) [reply solutions/cf-862B.mdx:147]:
```suggestion
	cout << c * uc - (n - 1) << "\n";
```

File: solutions/cf-862B.mdx [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+```cpp
+#include <iostream>
+#include <vector>
+using namespace std;
+
+const int MAX_N = 100000;
+
+// +1 for 1-indexed nodes
+vector<vector<int>> adj(MAX_N + 1);
```
Comment (bqi343) [inline solutions/cf-862B.mdx:118]: why not an array of vectors if you're giving it a fixed size?

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: solutions/cf-862B.mdx [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+<LanguageSection>
+
+<JavaSection>
+
+```java
+import java.util.*;
+import java.io.*;
+
+public class Main {
```
Comment (SansPapyrus683) [inline solutions/cf-862B.mdx:29]: nani why did you rename this to main

File: solutions/cf-862B.mdx [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+	public static ArrayList<Integer> adj[];
```
Comment (SansPapyrus683) [inline solutions/cf-862B.mdx:30]:
```suggestion
	public static ArrayList<Integer>[] adj;
```
this is how java does arrays

File: solutions/cf-862B.mdx [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+			// if the adjacent node isn't the parent,
+			// visit that node with the opposite color
+			if (u != parent) {
+				dfs(u, node, !color);
+			}
+		}
+	}
+}
+
+//BeginCodeSnip{Kattio}
```
Comment (SansPapyrus683) [inline solutions/cf-862B.mdx:77]: ryan i think something like
```
//CodeSnip{Kattio}
```
should work lol

File: solutions/cf-862B.mdx (hunk 6 above)
Comment (SansPapyrus683) [reply solutions/cf-862B.mdx:125]: wait why was this resolved

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
wait ryan alot of suggestions were resolved w/o a reason? if you reject in the future could you at least say why?

**Comment by ryanchou-dev** [comment]
> wait ryan alot of suggestions were resolved w/o a reason? if you reject in the future could you at least say why?

ahh! im so sorry, i thought i applied all of them :(
~~maybe i should stop doing prs right when i wake up~~

renamed to main because i was testing kattio locally, and i forgot to change it back

File: solutions/cf-862B.mdx (hunk 7 above)
Comment (ryanchou-dev) [reply solutions/cf-862B.mdx:118]: the answer exceeds int?

Comment (bqi343) [reply solutions/cf-862B.mdx:118]: That's why you made `c` and `uc` `long long`s, right?

File: solutions/cf-862B.mdx [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+	public static void main(String[] args) throws IOException{
+		Kattio io = new Kattio();
+
+		int n = io.nextInt();
+
+		// +1 for 1-indexed nodes
+		adj = new ArrayList[n + 1];
+		
+		for (int i = 0; i <= n; i++) {
+			adj[i] = new ArrayList<Integer>();
```
Comment (nathangong) [inline solutions/cf-862B.mdx:44]:
```suggestion
			adj[i] = new ArrayList<>();
```

File: solutions/cf-862B.mdx [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+			// if the adjacent node isn't the parent,
+			// visit that node with the opposite color
+			if (u != parent) {
+				dfs(u, node, !color);
+			}
+		}
+	}
+}
+
+//CodeSnip{Kattio}
```
Comment (nathangong) [inline solutions/cf-862B.mdx:77]: Kattio has to be inside the class
```suggestion
	}
	
	// CodeSnip{Kattio}
}
```

**Review by nathangong — CHANGES_REQUESTED** [review]
Nice, Java code is clean

File: solutions/cf-862B.mdx (hunk 7 above)
Comment (ryanchou-dev) [reply solutions/cf-862B.mdx:118]: oh mb, thanks!

**Review by bqi343 — APPROVED** [review]

**Review by Navigator365 — APPROVED** [review]
Good job!

File: solutions/cf-862B.mdx [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+const int MAX_N = 100000;
+
+// +1 for 1-indexed nodes
+vector<int> adj[MAX_N + 1];
+
+// counters for colored and uncolored nodes
+long long c = 0, uc = 0;
+
+void dfs(int node, int parent, bool col) {
+	(col ? c : uc)++;
```
Comment (dutinmeow) [inline solutions/cf-862B.mdx:96]: Write this explicitly with an `if` statement (no need to be so concise)

File: solutions/cf-862B.mdx [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+
+	for (int u : adj[node]) {
+		// if the adjacent node isn't the parent,
+		// visit that node with the opposite color
+		if (u != parent) dfs(u, node, !col);
```
Comment (dutinmeow) [inline solutions/cf-862B.mdx:101]: Put this on a newline with brackets

File: solutions/cf-862B.mdx [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+		
+		// ans = (all possible edges) - (existing edges)
+		long ans = (count[0] * count[1]) - (n - 1); 
+		
+		io.println(ans);
+		io.close();
+	}
+
+	public static void dfs(int node, int parent, boolean color) {
+		count[color ? 0 : 1]++;
```
Comment (dutinmeow) [inline solutions/cf-862B.mdx:65]: Use if statement here
```suggestion
		if (color) {
				count[0]++;
		}
		else {
				count[1]++;
		}
```

**Review by dutinmeow — CHANGES_REQUESTED** [review]
overall pretty good

File: solutions/cf-862B.mdx (hunk 16 above)
Comment (SansPapyrus683) [reply solutions/cf-862B.mdx:65]: dustin since when was a simple ternary statement so scary

Comment (dutinmeow) [reply solutions/cf-862B.mdx:65]: I feel like this is something you would do

Comment (SansPapyrus683) [reply solutions/cf-862B.mdx:65]: i mean this one is really simple, i only ask people to unwrap ungodly one-liners

File: solutions/cf-862B.mdx (hunk 14 above)
Comment (ryanchou-dev) [reply solutions/cf-862B.mdx:96]: i think it's fine...

File: solutions/cf-862B.mdx (hunk 16 above)
Comment (ryanchou-dev) [reply solutions/cf-862B.mdx:65]: also why are the indents two tabs?

**Review by nathangong — APPROVED** [review]

File: solutions/cf-862B.mdx (hunk 16 above)
Comment (dutinmeow) [reply solutions/cf-862B.mdx:65]: err thats my bad i suck at github

File: solutions/cf-862B.mdx (hunk 14 above)
Comment (dutinmeow) [reply solutions/cf-862B.mdx:96]: u know what its fine 
sans is getting into my head

**Review by dutinmeow — APPROVED** [review]

**Review by SansPapyrus683 — APPROVED** [review]
